<?xml version="1.0"?>
<TeamMentor_Article Metadata_Hash="217260072" Content_Hash="-713800716">
  <Metadata>
    <Id>09603b85-0c14-493e-b611-c10b60e5c6a9</Id>
    <Id_History>09603b85-0c14-493e-b611-c10b60e5c6a9,efa40bdd-f775-444b-af89-47e214fcc081,</Id_History>
    <Library_Id>be5273b1-d682-4361-99d9-6234f2d47eb7</Library_Id>
    <Title>Be Careful When Using Semaphores And Other Reference Counting Mechanisms</Title>
    <Category>Concurrency</Category>
    <Phase>Implementation</Phase>
    <Technology>Technology Independent</Technology>
    <Type>Guideline</Type>
    <DirectLink>Be Careful When Using Semaphores And Other Reference Counting Mechanisms</DirectLink>
    <Tag />
    <Security_Demand />
    <Author />
    <Priority />
    <Status />
    <Source>SI</Source>
  </Metadata>
  <Content Sanitized="true" DataType="markdown">
    <Data><![CDATA[## What to Do ##

Carefully track your acquires and releases when using a semaphore or other reference counting mechnisms to ensure the counting stays balanced.


## Why ##

If a semaphore becomes unbalanced it could either block system execution (too many acquires) or it could allow too many threads to access a resource (too many releases). Other reference counting mechanisms could cause other unexpected results such as memory leaks or release of interfaces before your code is done using them.


## When ##

When using a semaphore or any other reference counting mechanism.


## How ##

Trace your execution paths to make sure that each aquire is matched with a release. It is easiest to do this if the match is always made in the same function so that you don't have to trace between multiple functions or components. In the case of a semaphore it is best practice to use the resource for as little time as possible so matching a release closely to an acquire is a good idea. 

Be especially careful with error paths. When using exceptions place the release call in a finally block to ensure it is called in the course of normal execution or in the case of an exception. When using normal error handling logic, ensure that each return path out of a function releases the correct number of references.


## Problem Example ##

In the following example, the semaphore is acquired before the read call and is released after the `parseBuffer` call. However, it is not released in the case of an error with the read.

	bool ReadFromSocket(SOCKET socket, char *recvBuffer, int len)
	{
	  //first read from the socket
	  gSemaphore.Aquire();
	  int ret = read(socket, recvBuffer, len, 0);


	  if (ret >= 0)
	  {
		  //now parse the buffer
		  parseBuffer(recvBuffer, len);
		  gSemaphore.Release();
		  return true;
	  }
	  else
	  {
		  //can call WSAGetLastError if we need error details
		  return false;
	  }
	}

## Solution Example ##

In the following example, the semaphore is acquired before the read call and is released after the `parseBuffer` call and also in the error handler.

	bool ReadFromSocket(SOCKET socket, char *recvBuffer, int len)
	{
	  //first read from the socket
	  gSemaphore.Aquire();
	  int ret = read(socket, recvBuffer, len, 0);


	  if (ret >= 0)
	  {
		  //now parse the buffer
		  parseBuffer(recvBuffer, len);
		  gSemaphore.Release();
		  return true;
	  }
	  else
	  {
		  gSemaphore.Release();
		  return false;
	  }
	}]]></Data>
  </Content>
</TeamMentor_Article>